### ISSUE TITLE
Analyzers static ctor crashes with inability to read its own field

### ISSUE THREAD

**Issue by AArnott** [microsoft/VSSDK-Analyzers#105]
VSSDK001, 002, and 006 all failed simultaneously with roughly the same error.
It happened in an already opened solution.
vssdk-analyzers version: 16.3.2

```
System.TypeInitializationException : The type initializer for 'Microsoft.VisualStudio.SDK.Analyzers.VSSDK001DeriveFromAsyncPackageCodeFix' threw an exception. ---> System.FieldAccessException : Attempt by method 'Microsoft.VisualStudio.SDK.Analyzers.VSSDK001DeriveFromAsyncPackageCodeFix..cctor()' to access field 'Microsoft.VisualStudio.SDK.Analyzers.VSSDK001DeriveFromAsyncPackageAnalyzer.Descriptor' failed.
   at Microsoft.VisualStudio.SDK.Analyzers.VSSDK001DeriveFromAsyncPackageCodeFix..cctor()
   --- End of inner exception stack trace ---
   at Microsoft.VisualStudio.SDK.Analyzers.VSSDK001DeriveFromAsyncPackageCodeFix.get_FixableDiagnosticIds()
   at Microsoft.CodeAnalysis.CodeFixes.CodeFixService.GetAndTestFixableDiagnosticIds(CodeFixProvider codeFixProvider)
   at Microsoft.CodeAnalysis.CodeFixes.CodeFixService.<>c.<GetFixableDiagnosticIds>b__27_2(CodeFixProvider f)
   at System.Collections.Immutable.ImmutableInterlocked.GetOrAdd[TKey,TValue](ImmutableDictionary`2& location,TKey key,Func`2 valueFactory)
   at Microsoft.CodeAnalysis.CodeFixes.CodeFixService.<>c__DisplayClass27_0.<GetFixableDiagnosticIds>b__0()
   at Microsoft.CodeAnalysis.Extensions.IExtensionManagerExtensions.PerformFunction[T](IExtensionManager extensionManager,Object extension,Func`1 function,T defaultValue)
```

**Comment by japj** [comment]
Could it be that this is due to "Internals" of the Analyzer is not visible to the CodeFix Assembly?
(I am not entirely sure, but it looks there was some reshuffling in project/assembly boundaries in the past?) not sure why the unit tests pass though

**Comment by AArnott** [comment]
The issue here is already understood: when a solution references multiple versions of the analyzer package, or when projects are upgraded from one version to another, the analyzer assembly will be loaded with one version and the code fix assembly will be loaded with another. These two assemblies are designed to work together, so having divergent versions loaded is a breaking configuration.

But I feel like https://github.com/microsoft/VSSDK-Analyzers/commit/e69a0c3ebdc9aff9a4563e2be750db83128aa4f3 would have prevented this problem.

> vssdk-analyzers version: 16.0.29 is also broken

You mean with this error, or something else?

**Comment by japj** [comment]
> > vssdk-analyzers version: 16.0.29 is also broken
> 
> You mean with this error, or something else?

Yes, with this error.

I am using CentralPackageVersioning with
```xml
      <PackageReference Update="Microsoft.VisualStudio.SDK" Version="16.0.202" />
      <PackageReference Update="Microsoft.VSSDK.BuildTools" Version="16.4.12" />
```

As far as I understand `Microsoft.VisualStudio.SDK` 16.0.202 takes a (minimal) dependency on Analyzers 16.3.2

I tried downgrading to older versions of the Analyzers (and SDK: 16.0.200 and lower) in order to get to a point where it would work again, but even if I do a `git clean -dfx` on the plugin repo and open the solution I still end up with this issue.

Thus at the moment, the analyzers seem to be unusable for me.
I was hoping that a workaround would be to downgrade to a specific version, but that does not seem to be the case.

Note this is with VS2019 16.2.2

So I am a bit at a loss to understand what exactly is causing this issue to persist in our case

**Comment by AArnott** [comment]
Thanks for sharing details. I'm motivated to help you resolve this and I hope to learn more about the problem as well, since if you're always using one consistent version I would not expect you to hit this problem.
Before we investigate your current setup though, can you try the [latest analyzer version](https://www.nuget.org/packages/microsoft.visualstudio.sdk.analyzers) (16.3.14)?

**Comment by japj** [comment]
I cleaned up my nuget package cache and did a clean nuget restore and discovered that multiple different versions of the analyzers were restored (this is a solution that also includes some project wizards/templates next to the plugin).

I explicitly added the 16.3.14 analyzers to the ProjectReferences of each projects and did the same: now only one nuget package version was in the cache.

So I think the problem can probably be reproduced by having 2 VSSDK projects in the same solution that trigger different nuget versions of the analyzers (like you mentioned earlier).

Hope this helps?

**Comment by AArnott** [comment]
Yes, that confirms my theory. Thank you.

I just wish I knew why having precise assembly versions doesn't fix it. If you could tell me which exact package versions showed up in your package cache that could help me confirm whether the precise assembly version fix I made a while back was *not* included in any of them.

**Comment by japj** [comment]
These are the versions (and their "parent" nuget that takes the dependency:

- Microsoft.VSSDK.BuildTools 16.4.12 depends on "Microsoft.VisualStudio.SDK.Analyzers" version="15.8.33"

- Microsoft.VisualStudio.SDK 16.0.202 depends on "Microsoft.VisualStudio.SDK.Analyzers" version="16.3.2"

**Comment by AArnott** [comment]
Thanks. I hadn't realized that `Microsoft.VSSDK.BuildTools` had a dependency, and such an old one at that. I'll get this fixed in the next release of that package.

**Comment by AArnott** [comment]
It turns out we're not using strong names in our code fix and analyzer assemblies. Without strong names, the desktop CLR won't allow loading the assembly multiple times (once per assembly version). That's what's causing this.

**Comment by AArnott** [comment]
Fixed in 16.5.5
